edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/NotificationHandler.cs

```diff
@@ -0,0 +1,152 @@
+            Func<T, Task> notificationStorer = null,
+            Func<Task<IEnumerable<Message>>> storedNotificationRetriever = null)
+        {
+            this.notificationConvertor = Preconditions.CheckNotNull(notificationConvertor);
+            this.notificationStorer = notificationStorer;
+            this.storedNotificationRetriever = storedNotificationRetriever;
+            this.connected = false;
+        }
+
+        public void SetConnector(IMqttBrokerConnector connector)
+        {
+            this.connector = Preconditions.CheckNotNull(connector);
+            this.connector.OnConnected += async (sender, args) => await this.OnConnectAsync();
+        }
+
+        async Task OnConnectAsync()
+        {
+            using (await this.stateLock.LockAsync())
+            {
+                this.connected = true;
+                if (this.storedNotificationRetriever != null)
+                {
+                    var messages = await this.storedNotificationRetriever();
+                    await this.SendMessagesAsync(messages);
+                }
+            }
+        }
+
+        public async Task NotifyAsync(T notification)
+        {
+            using (await this.stateLock.LockAsync())
+            {
+                if (this.connected)
+                {
+                    var messages = await this.notificationConvertor(notification);
+                    await this.SendMessagesAsync(messages);
+                }
+                else if (this.notificationStorer == null)
+                {
+                    Events<T>.NotificationDiscarded();
+                }
+                else
+                {
+                    Events<T>.StoringNotification();
+                    await this.notificationStorer(notification);
+                    Events<T>.NotificationStored();
+                }
+            }
+        }
+
+        async Task SendMessagesAsync(IEnumerable<Message> messages)
+        {
+            if (messages != null)
+            {
+                var tasks = new List<Task>();
+                foreach (var message in messages)
+                {
+                    tasks.Add(this.SendMessageAsync(message));
+                }
+
+                await Task.WhenAll(tasks);
+            }
+        }
+
+        async Task SendMessageAsync(Message message)
+        {
+            Events<T>.SendingNotification(message);
+            try
+            {
+                var delivered = await this.connector.SendAsync(message.Topic, Encoding.UTF8.GetBytes(message.Payload));
+                if (delivered)
+                {
+                    Events<T>.NotificationSent(message);
+                }
+                else
+                {
+                    Events<T>.NotificationNotDelivered(message);
+                }
+            }
+            catch (Exception ex)
+            {
+                Events<T>.SendingNotificationError(message, ex);
+            }
+        }
+    }
+
+    public class Message
```

Comment:
Should we rename this BrokerMessage? Message is already overloaded in EdgeHub

Reply:
done